### ISSUE TITLE
[🐞] Can't build project via `bun run --bun`

### ISSUE THREAD

**Issue by octet-stream** [QwikDev/qwik#5917]
### Which component is affected?

Qwik Rollup / Vite plugin

### Describe the bug

I've been trying to build a project using Bun's [feature](https://bun.sh/docs/cli/run#bun) that allows to replace Node.js binary for a command. However, while Qwik has adapter for Bun's http server, it does not seem to have a full support for bun, because when I'm trying to build my project like this I get following error:

```
$ bun run --bun tsc --incremental --noEmit --pretty
$ bun run --bun vite build
vite v5.1.4 building for production...
✓ 0 modules transformed.
x Build failed in 4ms
error during build:
Error: [vite-plugin-qwik] Not implemented
    at <anonymous> (/Users/octetstream/projects/qwik-bun-build-issue/node_modules/@builder.io/qwik/optimizer.mjs:1435:35)
    at processTicksAndRejections (:12:39)
error: "vite" exited with code 1
error: script "build.client" exited with code 1
error: "qwik" exited with code 1
error: script "build" exited with code 1
```

The error is pretty clear - you don't seem to support bun runtime in qwik optimizer. I've been trying to add bun to `getPlatformInputFiles`, just to see it that would be enough (because Bun aims to be compatible with Node.js), and it seems like it is not. When I do so, a new error occurs:

```
$ bun run --bun tsc --incremental --noEmit --pretty
$ bun run --bun vite build
vite v5.1.4 building for production...
✓ 0 modules transformed.
x Build failed in 6ms
error during build:
undefined
error: "vite" exited with code 1
error: script "build.client" exited with code 1
error: "qwik" exited with code 1
error: script "build" exited with code 1
```

### Reproduction

https://github.com/octet-stream/qwik-bun-build-issue

### Steps to reproduce

1. Clone this repository via git clone `git@github.com:octet-stream/qwik-bun-build-issue.git`;
2. Install dependencis via `bun i`;
3. Run `bun run build` and the build process with fail with the error.

### System Info

```shell
System:
    OS: macOS 14.3.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 98.50 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.6.2 - ~/Library/Caches/fnm_multishells/2008_1708258753953/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 10.2.4 - ~/Library/Caches/fnm_multishells/2008_1708258753953/bin/npm
    pnpm: 8.15.4 - /opt/homebrew/bin/pnpm
    bun: 1.0.29 - /opt/homebrew/bin/bun
  Browsers:
    Chrome: 122.0.6261.69
    Safari: 17.3.1
  npmPackages:
    @builder.io/qwik: ^1.4.5 => 1.4.5 
    @builder.io/qwik-city: ^1.4.5 => 1.4.5 
    undici: * => 6.6.2 
    vite: ^5.0.12 => 5.1.4
```

### Additional Information

The current solution is to just keep using Node.js to build Qwik City apps. This also means Node.js should be presented during the build stage when deploying, which is not a huge problem when using Docker multilayer builds, I think. But even with Docker that would mean an unnecessary Node.js setup, just to build a project.

**Comment by gioboa** [comment]
I see. Did you try with an older version of Qwik? I'm quite sure it worked fine.

**Comment by octet-stream** [comment]
Tested with older versions after your comment in separate repository, because the simplest way to get appropriate dependencies is to run `bun create qwik@<version>`, which is dropping non-empty directory including `.git`. This error occurs even in https://github.com/BuilderIO/qwik/releases/tag/v1.2.12 which is seem the one introducing bun adapter. Are you sure you have run build with `bun run --bun`? This flag is necessary to replace Node.js' binary with Bun's.

Try with 1.2.12:

1. Create qwik city app via: `bun create qwik@1.2.12`
2. Run `bun run qwik add bun`
3. Then create client bundle via `bun run --bun build.client`

**Comment by Polygons1** [comment]
I have the same issue and looks like the function `getPlatformInputFiles` in the optimizer doesn't check for bun and only for node before returning null

**Comment by octet-stream** [comment]
Yes, but then if you just add bun to that condition, it will just fail with another error. And this error just says something like `error: undefined`.

**Comment by Polygons1** [comment]
I just removed the condition for bun, and it worked

**Comment by octet-stream** [comment]
Can confirm this workaround helps. But in my case the build freezes during SSG, and also fails if `TextEncoderStream` is not defined, which is the case for Bun. See https://github.com/QwikDev/qwik/issues/5929 for more info and https://github.com/QwikDev/qwik/pull/6310#issuecomment-2116414641 for a polyfill (if at the moment you read this comment Qwik has not been released a new version including this polyfill)

A better way to fix it would be to look where `getEnv` is used across the codebase and add checks for Bun. But for now it might be enough, since Bun aims for Node.js compatibility. I can send a PR once I figure out why `build.server` freezes.

**Comment by octet-stream** [comment]
Ah, sometimes it says `worker exit 0` and just freezes during server build. The client build and tsc check finished without any problem. So, I wonder what happens during server build.

**Comment by PatrickJS** [comment]
@octet-stream you can also add the polyfill in vite.config so it's available during build to see

**Comment by octet-stream** [comment]
Sure, but first I'll figure out why `build.server` command hangs. From what I can understand, the problem might occur due to some unknown issue. I can try and isolate the issue in my repro repository, if somebody can help me figure out the source of the problem.

**Comment by octet-stream** [comment]
Yes, I can confirm the `build.server` command hangs with `bun run --bun`. Updated my [repro](https://github.com/octet-stream/qwik-bun-build-issue). I won't open a separate issue, because they're related and one occurs when the other is fixed, and if I find the source of this problem, I might be able to fix both in one PR.

How to get to this issue:
1. Clone the repository
2. Install dependencies via `bun install`
3. Open `node_modules/@builder.io/qwik/optimizer.mjs`
4. Comment out lines 1405...1407
5. Run `bun run --bun build.server`

The compiler freezes during SSG. To confirm this is a SSG issue, add `{ssg: null}` to `bunServerAdapter` call in `adapters/bun/vite.config.ts` file, then run `bun run --bun build.server` command. Now everything works as expected.

As I said before, sometimes `worker exit 0` error message shows up in the console, but sometimes it just freezes silently. I am currently looking to `node_modules/@builder.io/qwik-city/static/node.mjs` as potential source of the issue.

**Comment by octet-stream** [comment]
For some reason the "message" event [here](https://github.com/QwikDev/qwik/blob/a2f37398057318dd6f99f8a9e763a876cc0c89a4/packages/qwik-city/static/node/node-worker.ts#L7-L9) is never fired in Bun. This can be a bug in Bun implementation for `node:worker_threads`.

**Comment by PatrickJS** [comment]
yeah, might be

**Comment by octet-stream** [comment]
My thought were that Bun can't send object through `postMessage` method pack and forth, but it's not the case. It does work as expected in the simplest worker thread setup (I can send and receive objects in main thread and worker). Yet during Qwik's SSG the process hangs and I don't see "message" listeners to receive messages on both sides. I have no idea what is going on here.

This is the line on which I think the build process hangs: https://github.com/QwikDev/qwik/blob/a2f37398057318dd6f99f8a9e763a876cc0c89a4/packages/qwik-city/static/node/node-main.ts#L117

**Comment by PatrickJS** [comment]
might be async issue with bun where a worker was setup after the message was sent already. maybe you can add some retry setTimeout logic after a little but of time and limited retries

**Comment by octet-stream** [comment]
There's definitely something going on with Worker threads in Bun, maybe I should report an issue to them, I'm not sure. When I set `maxWorkers` to 1 in `ssg` option everything works as expected, so the problem is here somewhere. And they do for sure have different behavior, like I just found out in my minimal workers example that Node.js throws an error when `Worker` is constructed with `import.meta.url` saying that it should be and absolute path and the `file://` urls must be wrapped to the `URL` object, and then Bun just doesn't care. So maybe it's worth to open an issue in Bun's repository, since they're trying to be compatible with Node. But then again, they unfamiliar with Qwik codebase, so it may take some effort to figure things out for them.

Anyway, I'm gonna put together a PR just to enable Bun support for app builds later today.
